### PR TITLE
SI-8177 specializeSym must use memberInfo on high side

### PR DESCRIPTION
**Pull request by adriaanm**
When determining whether member `symLo` of `tpLo`
has a stronger type than member `symHi` of `tpHi`,
should we use memberType or memberInfo?

Well, memberType transforms (using `asSeenFrom`) `sym.tpe`,
whereas memberInfo performs the same transform on `sym.info`.
For term symbols, this ends up being the same thing (`sym.tpe == sym.info`).
For type symbols, however, the `.info` of an abstract type member
is defined by its bounds, whereas its `.tpe` is a `TypeRef` to that type symbol,
so that `sym.tpe <:< sym.info`, but not the other way around.

Thus, for the strongest (correct) result,
we should use `memberType` on the low side.

On the high side, we should use the result appropriate
for the right side of the `<:<` above (`memberInfo`).

### CODE REVIEW

**Comment by adriaanm** [comment]
review by @retronym, who discovered the flaw

**Comment by adriaanm** [comment]
working on the test failures

**Comment by retronym** [comment]
LGTM (assuming the "no type tag" test failure doesn't unveil another layer of pain...)

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -4100,23 +4100,43 @@ trait Types
...
+      isSubType(tpLo, tpHi, depth)        &&
+      (!symHi.isStable || symLo.isStable) &&               // sub-member must remain stable
+      (!symLo.hasVolatileType || symHi.hasVolatileType)    // sub-member must not introduce volatility
+    } else if (symHi.isAbstractType) {
+      (tpHi.bounds containsType tpLo) &&
+      kindsConform(symHi :: Nil, tpLo :: Nil, preLo, symLo.owner)
+    } else {
+      // class symbols can't be specialized
+      (symHi.isAliasType) &&  tpLo =:= tpHi
+    }
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala:4140]: Wow, so _that's_ what it is meant to do!

File: test/files/neg/t8177a.scala [hunk 2]
```diff
@@ -0,0 +1,6 @@
+trait A { type Result }
+
+class PolyTests {
+  def wrong(x: A { type Result = Int })
+             : A { type Result = String} = x
+}
```
Comment (retronym) [inline test/files/neg/t8177a.scala:6]: It's mind boggling that we didn't already have test coverage of this, no?

File: src/reflect/scala/reflect/internal/Types.scala (hunk 1 above)
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala:4140]: I hope so, but the test failure is pretty painful to dissect. The new way looks right to me, but I'm  definitely doing some more poking before I feel confident enough...

File: test/files/neg/t8177a.scala (hunk 2 above)
Comment (adriaanm) [inline test/files/neg/t8177a.scala:6]: Yep. Will make for great dotty test cases too.

File: src/reflect/scala/reflect/internal/Types.scala (hunk 1 above)
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala:4140]: The failure is because now we dealias less aggressively and the tag for the type alias can't be materialized. Looks like a problem there rather than in my fix. /cc @xeno-by

Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala:4140]: Yep, `typeTag[<refinementClass].T]` does not work -- before we dealiased the selection away. Added a backstop.

**Comment by adriaanm** [comment]
tests should pass now

File: src/compiler/scala/reflect/reify/codegen/GenTypes.scala [hunk 3]
```diff
@@ -55,7 +55,9 @@ trait GenTypes {
       case tpe @ ConstantType(value) =>
         mirrorFactoryCall(nme.ConstantType, reifyProduct(value))
       case tpe @ TypeRef(pre, sym, args) =>
-        reifyProduct(tpe)
+        // TODO: remove special case!!! for now, as we can't reify these symbols, let's hope dealias gets us out of this bind...
```
Comment (adriaanm) [inline src/compiler/scala/reflect/reify/codegen/GenTypes.scala]: /cc @xeno-by

**Comment by adriaanm** [comment]
Well, _now_ they should pass. Let's please try to void uncommented foo/bar/baz tests in the future.

File: test/files/run/t6992/Test_2.scala [hunk 4]
```diff
@@ -2,7 +2,7 @@ import scala.language.reflectiveCalls
 
 object Test extends App {
   val foo = Macros.foo("T")
-  println(scala.reflect.runtime.universe.weakTypeOf[foo.T].typeSymbol.typeSignature)
+  println(scala.reflect.runtime.universe.weakTypeOf[foo.T])
```
Comment (xeno-by) [inline test/files/run/t6992/Test_2.scala]: Why this change?

File: src/compiler/scala/reflect/reify/codegen/GenTypes.scala (hunk 3 above)
Comment (xeno-by) [inline src/compiler/scala/reflect/reify/codegen/GenTypes.scala]: Why do we need this special case? Could you elaborate in a bit more detail?

Comment (xeno-by) [inline src/compiler/scala/reflect/reify/codegen/GenTypes.scala]: Okay saw the above commit.

File: src/reflect/scala/reflect/internal/Types.scala (hunk 1 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Types.scala:4140]: Could you explain how I can reproduce this? Is it SI-6992?

File: test/files/run/t6992/Test_2.scala (hunk 4 above)
Comment (xeno-by) [inline test/files/run/t6992/Test_2.scala]: Okay now I see why. How about actually testing both?

File: src/reflect/scala/reflect/internal/Types.scala (hunk 1 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Types.scala:4140]: Was it SI-8104? If yes, then TypeTag should be changed to WeakTypeTag, and everything should work.

**Comment by xeno-by** [comment]
Here's the proposed diff that doesn't require specialcasing `GenTypes`.

``` diff
09:38 ~/Projects/8177/sandbox (t8177b)$ git diff
diff --git a/src/compiler/scala/reflect/reify/codegen/GenTypes.scala b/src/compiler/scala/reflect/reify/codegen/GenTypes.scala
index 3e2acc28e5..a90a3a338b 100644
--- a/src/compiler/scala/reflect/reify/codegen/GenTypes.scala
+++ b/src/compiler/scala/reflect/reify/codegen/GenTypes.scala
@@ -55,9 +55,7 @@ trait GenTypes {
       case tpe @ ConstantType(value) =>
         mirrorFactoryCall(nme.ConstantType, reifyProduct(value))
       case tpe @ TypeRef(pre, sym, args) =>
-        // TODO: remove special case!!! for now, as we can't reify these symbols, let's hope dealias gets us out of this bind...
-        if (pre.typeSymbol.isAnonOrRefinementClass && (tpe ne tpe.dealias)) reifyType(tpe.dealias)
-        else reifyProduct(tpe)
+        reifyProduct(tpe)
       case tpe @ TypeBounds(lo, hi) =>
         reifyProduct(tpe)
       case tpe @ NullaryMethodType(restpe) =>
diff --git a/src/reflect/scala/reflect/internal/Types.scala b/src/reflect/scala/reflect/internal/Types.scala
index 27ec882549..b2e36a2732 100644
--- a/src/reflect/scala/reflect/internal/Types.scala
+++ b/src/reflect/scala/reflect/internal/Types.scala
@@ -1143,7 +1143,7 @@ trait Types
   /** A class for this-types of the form <sym>.this.type
    */
   abstract case class ThisType(sym: Symbol) extends SingletonType with ThisTypeApi {
-    if (!sym.isClass) {
+    if (!sym.isClass && !sym.isFreeType) {
       // SI-6640 allow StubSymbols to reveal what's missing from the classpath before we trip the assertion.
       sym.failIfStub()
       abort(s"ThisType($sym) for sym which is not a class")
diff --git a/test/files/neg/t8104/Test_2.scala b/test/files/neg/t8104/Test_2.scala
index 585f76c00f..a3bd940188 100644
--- a/test/files/neg/t8104/Test_2.scala
+++ b/test/files/neg/t8104/Test_2.scala
@@ -9,7 +9,7 @@ object Test extends App {
   case class C(x: Int, y: Int)

   import scala.reflect.runtime.universe._
-  def reprify[T, Repr](x: T)(implicit generic: Generic.Aux[T, Repr], tag: TypeTag[Repr]) = println(tag)
+  def reprify[T, Repr](x: T)(implicit generic: Generic.Aux[T, Repr], tag: WeakTypeTag[Repr]) = println(tag)
   reprify(C(40, 2))

   // this is a compilation error at the moment as explained in SI-8104
diff --git a/test/files/run/t6992.check b/test/files/run/t6992.check
index 1a0684c995..021f32ec95 100644
--- a/test/files/run/t6992.check
+++ b/test/files/run/t6992.check
@@ -1,3 +1,4 @@
+Test.foo.T
 Int
 42
 42
diff --git a/test/files/run/t6992/Test_2.scala b/test/files/run/t6992/Test_2.scala
index 0f226b01ce..1ed8958d38 100644
--- a/test/files/run/t6992/Test_2.scala
+++ b/test/files/run/t6992/Test_2.scala
@@ -2,7 +2,9 @@ import scala.language.reflectiveCalls

 object Test extends App {
   val foo = Macros.foo("T")
-  println(scala.reflect.runtime.universe.weakTypeOf[foo.T])
+  val ttpe = scala.reflect.runtime.universe.weakTypeOf[foo.T]
+  println(ttpe)
+  println(ttpe.typeSymbol.typeSignature)

   val bar = Macros.bar("test")
   println(bar.test)
diff --git a/test/files/run/t8104.check b/test/files/run/t8104.check
index c2593eb199..40523a2868 100644
--- a/test/files/run/t8104.check
+++ b/test/files/run/t8104.check
@@ -1 +1,2 @@
-TypeTag[(Int, Int)]
+WeakTypeTag[<refinement>.this.Repr]
+(Int, Int)
diff --git a/test/files/run/t8104/Test_2.scala b/test/files/run/t8104/Test_2.scala
index 630176f175..55c080a563 100644
--- a/test/files/run/t8104/Test_2.scala
+++ b/test/files/run/t8104/Test_2.scala
@@ -9,7 +9,10 @@ object Test extends App {
   case class C(x: Int, y: Int)

   import scala.reflect.runtime.universe._
-  def reprify[T, Repr](x: T)(implicit generic: Generic.Aux[T, Repr], tag: TypeTag[Repr]) = println(tag)
+  def reprify[T, Repr](x: T)(implicit generic: Generic.Aux[T, Repr], tag: WeakTypeTag[Repr]) = {
+    println(tag)
+    println(tag.tpe.typeSymbol.typeSignature)
+  }
   reprify(C(40, 2))

   implicitly[Generic.Aux[C, (Int, Int)]]
```

**Comment by adriaanm** [comment]
thanks, incorporated your patch
